### PR TITLE
make memcmp return a value of c_int_width instead of i32

### LINKED ISSUE #32610: core dependency on c_int in memcmp
libcore/str/mod.rs uses `memcmp`, and declares it to return i32, in reality it returns c_int, so this is a potential portability concern.

Maybe memcmp can be made an intrinsic to resolve this.

### PR DESCRIPTION
**Pull request by drmorr0**
This is an attempt to fix #32610 and #78022, namely, that `memcmp` always returns an `i32` regardless of the platform.  I'm running into some issues and was hoping I could get some help.

Here's what I've been attempting so far:

1. Build the stage0 compiler with all the changes _expect_ for the changes in `library/core/src/slice/cmp.rs` and `compiler/rustc_codegen_llvm/src/context.rs`; this is because `target_c_int_width` isn't passed through and recognized as a valid config option yet.  I'm building with `./x.py build --stage 0 library/core library/proc_macro compiler/rustc`
2. Next I add in the `#[cfg(c_int_width = ...)]` params to `cmp.rs` and `context.rs` and build the stage 1 compiler by running `./x.py build --keep-stage 0 --stage 1 library/core library/proc_macro compiler/rustc`.  This step now runs successfully.
3. Lastly, I try to build the test program for AVR mentioned in #78022 with `RUSTFLAGS="--emit llvm-ir" cargo build --release`, and look at the resulting llvm IR, which still shows:

```
...
%11 = call addrspace(1) i32 @memcmp(i8* nonnull %5, i8* nonnull %10, i16 5) #7, !dbg !1191                                                                                                                                                                                                                                %.not = icmp eq i32 %11, 0, !dbg !1191
...
; Function Attrs: nounwind optsize                                                                                                                                                                                                                                                                                          declare i32 @memcmp(i8*, i8*, i16) local_unnamed_addr addrspace(1) #4
```

Any ideas what I'm missing here?  Alternately, if this is totally the wrong approach I'm open to other suggestions.

cc @Rahix

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/context.rs [hunk 1]
```diff
@@ -777,6 +777,9 @@ impl CodegenCx<'b, 'tcx> {
 
         // This isn't an "LLVM intrinsic", but LLVM's optimization passes
         // recognize it like one and we assume it exists in `core::slice::cmp`
+        #[cfg(c_int_width = "16")]
+        ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16);
+        #[cfg(c_int_width = "32")]
         ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i32);
```
Comment (Rahix) [inline compiler/rustc_codegen_llvm/src/context.rs:783]: I don't think this is right.  This code is part of the compiler and as such your `cfg()`s apply to the build architecture, not the one you're compiling for.  I think this needs to be something like

```rust
match self.sess().target.c_int_width  {
    16 => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16),
    32 => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i32),
    _ => /* what to do as the fallback? */,
}
```

Comment (joshtriplett) [reply compiler/rustc_codegen_llvm/src/context.rs:783]: Also, there exist platforms with a 64-bit c_int (ILP64 platforms), though I don't think we currently support any.

Comment (drmorr0) [reply compiler/rustc_codegen_llvm/src/context.rs:783]: Oh!  Yes.  This makes perfect sense.  Thank you!  I'll see if I can fix this up today.

**Comment by drmorr0** [comment]
Ok, thanks to your pointers I got this actually working and confirmed that this works for the test program.

I _think_ I need to do this in two parts to pass CI?  The first part will make the compiler handle different c_int_widths, and then I can uncomment the `#[cfg(...)]` bits in core.

Also, I'm unsure what to do for the "default" case for the match statement... currently this outputs the following if you use a bad value for `target-c-int-width`:

```
error: internal compiler error: compiler/rustc_codegen_llvm/src/context.rs:784:18: Unsupported target_c_int_width = 17                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1147:9                                                                                                                                                                                                                                              note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   note: the compiler unexpectedly panicked. this is a bug.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        note: rustc 1.58.0-dev running on x86_64-unknown-linux-gnu                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      note: compiler flags: -Z unstable-options -C opt-level=s -C panic=abort -C linker-plugin-lto -C codegen-units=1 -C debuginfo=2 --crate-type lib                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 note: some of the compiler flags provided by cargo are hidden
```

Lastly, are there any tests I should write, or any other things I can do to help get this through?

**Comment by cuviper** [comment]
> I _think_ I need to do this in two parts to pass CI? The first part will make the compiler handle different c_int_widths, and then I can uncomment the `#[cfg(...)]` bits in core.

You can use `cfg(bootstrap)` to work with the stage0 compiler that doesn't have your changes yet.
e.g. combined: `#[cfg(any(bootstrap, c_int_width = "32"))]`

**Comment by drmorr0** [comment]
> You can use `cfg(bootstrap)` to work with the stage0 compiler that doesn't have your changes yet.
> e.g. combined: `#[cfg(any(bootstrap, c_int_width = "32"))]`

Fixed, thanks!

File: compiler/rustc_codegen_llvm/src/context.rs [hunk 2]
```diff
@@ -777,7 +777,12 @@ impl CodegenCx<'b, 'tcx> {
 
         // This isn't an "LLVM intrinsic", but LLVM's optimization passes
         // recognize it like one and we assume it exists in `core::slice::cmp`
-        ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i32);
+        match self.sess().target.c_int_width.as_str() {
+            "16" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16),
+            "32" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i32),
+            "64" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i64),
+            x => bug!("Unsupported target_c_int_width = {}", x),
```
Comment (scottmcm) [inline compiler/rustc_codegen_llvm/src/context.rs:784]: I came in here to mention this use; kudos for catching it already! 👍

**Comment by drmorr0** [comment]
@joshtriplett Hi, just checking if there's anything else I need to do on this to get it merged in?

File: library/core/src/slice/cmp.rs [hunk 3]
```diff
@@ -7,15 +7,20 @@ use crate::mem;
...
-    /// Returns 0 for equal, < 0 for less than and > 0 for greater
-    /// than.
-    // FIXME(#32610): Return type should be c_int
+    #[cfg(c_int_width = "16")]
+    fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i16;
+    // Assume we're building the bootstrap compiler on a system with 32-bit c-int width
+    #[cfg(any(bootstrap, c_int_width = "32"))]
     fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i32;
+    #[cfg(c_int_width = "64")]
+    fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i64;
```
Comment (petrochenkov) [inline library/core/src/slice/cmp.rs:23]: I don't think it's necessary to expand the compiler's public interface for this (introduce `#[cfg(int_width = ...)]`), `type c_int` can be defined just by `cfg`ing on target architectures, like it's done in `library\std\src\os\raw\mod.rs` or in the `libc` crate.

Right now there are only two architectures that differ from the `i32` default - `avr` and `msp430`.

File: compiler/rustc_codegen_gcc/src/common.rs [hunk 4]
```diff
@@ -317,47 +322,37 @@ impl<'gcc, 'tcx> SignType<'gcc, 'tcx> for Type<'gcc> {
...
-        else if self.is_i64(cx) {
+        } else if self.is_i64(cx) {
             cx.u64_type
-        }
-        else if self.is_i128(cx) {
+        } else if self.is_i128(cx) {
             cx.u128_type
-        }
-        else {
+        } else {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_gcc/src/common.rs:349]: Could you remove unrelated formatting changes from this file?

**Comment by petrochenkov** [comment]
The codegen changes look reasonable to me, although I'm not an expert in that area.

**Comment by mrk-its** [comment]
I'm testing these changes in my rust fork for mos arch (8bit 6502 CPU) and everything works fine.

**Review by nagisa — CHANGES_REQUESTED** [review]
I'm going to request that Vadim's review comments are addressed. I don't believe this PR should land as is.

**Comment by Dylan-DPC** [comment]
@drmorr0 any updates on this?

**Comment by drmorr0** [comment]
@Dylan-DPC I have not had a chance to revisit this yet, I'm hoping to get to it soon-ish, though.

File: library/core/src/slice/cmp.rs [hunk 5]
```diff
@@ -13,7 +13,9 @@ extern "C" {
     ///
     /// Returns 0 for equal, < 0 for less than and > 0 for greater
     /// than.
-    // FIXME(#32610): Return type should be c_int
+    #[cfg(any(target_arch = "avr", target_arch = "msp430",))]
+    fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i16;
+    #[cfg(not(any(target_arch = "avr", target_arch = "msp430",)))]
     fn memcmp(s1: *const u8, s2: *const u8, n: usize) -> i32;
```
Comment (petrochenkov) [inline library/core/src/slice/cmp.rs:19]: We actually have `c_int` in libcore now - https://doc.rust-lang.org/nightly/core/ffi/type.c_int.html
Although it's not defined correctly for AVR and MSP430.

So the correct definition of `c_int` can go there now, and `memcmp` can simply use it.

File: compiler/rustc_codegen_llvm/src/context.rs [hunk 6]
```diff
@@ -859,7 +859,11 @@ impl<'ll> CodegenCx<'ll, '_> {
 
         // This isn't an "LLVM intrinsic", but LLVM's optimization passes
         // recognize it like one and we assume it exists in `core::slice::cmp`
-        ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i32);
+        match self.sess().target.arch.as_str() {
+            "avr" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16),
```
Comment (petrochenkov) [inline compiler/rustc_codegen_llvm/src/context.rs:863]:
```suggestion
            "avr" | "msp430" => ifn!("memcmp", fn(i8p, i8p, t_isize) -> t_i16),
```
(And in `intrinsic.rs` too.)

File: library/core/src/slice/cmp.rs (hunk 5 above)
Comment (drmorr0) [reply library/core/src/slice/cmp.rs]: Neat!  Thanks

**Comment by drmorr0** [comment]
Ok I _think_ this is ready; I'm trying to get my test environment back up so that I can confirm it's generating the right thing but some other stuff has broken in the interim.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+
